### ISSUE TITLE
dart:ui types not handled by mobx_codegen properly on Flutter stable/beta

### ISSUE THREAD

**Issue by Bwolfs2** [mobxjs/mobx.dart#381]
Some people in the community  are reporting to me that the Color class when placed in Observable is generating Dynamic.

``` 
  @observable
  Color colorPrimary = null;
````

And generated it:

```
  @override
  dynamic get colorPrimary {
    _$colorPrimaryAtom.context.enforceReadPolicy(_$colorPrimaryAtom);
    _$colorPrimaryAtom.reportObserved();
    return super.colorPrimary;
  }
```

The Color class is blocked somehow and we are unable to solve it, has anyone had this problem?

**Comment by irvine5k** [comment]
I have this problem too.

**Comment by shyndman** [comment]
This should work.

Can you please verify that you're running the latest version of mobx_codegen? What's version does your pubspec.lock say it's using?

**Comment by irvine5k** [comment]
```
mobx_codegen: ^0.4.0+1
build_runner: ^1.7.3


 mobx_codegen:
    dependency: "direct dev"
    description:
      name: mobx_codegen
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.4.0+1"
```

When generate files throws this error on Class: 

`'_$Cartouche.backColor' ('List<dynamic> Function()') isn't a valid override of '_Cartouche.backColor' ('List<Color> Function()').dart(invalid_override)`

This is the store class: 

```
import 'package:flutter/material.dart';
import 'package:mobx/mobx.dart';

part 'cartouche.g.dart';

class Cartouche = _Cartouche with _$Cartouche;

abstract class _Cartouche with Store {

  @observable
  List<bool> color = ObservableList.of(List.generate(60, (i) => false));

  @observable
  List<Color> backColor = ObservableList.of(List.generate(60, (i) => Colors.transparent));

}

```

Generated code:

```
@override
  List<dynamic> get backColor {
    _$backColorAtom.context.enforceReadPolicy(_$backColorAtom);
    _$backColorAtom.reportObserved();
    return super.backColor;
  }

  @override
  set backColor(List<dynamic> value) {
    _$backColorAtom.context.conditionallyRunInAction(() {
      super.backColor = value;
      _$backColorAtom.reportChanged();
    }, _$backColorAtom, name: '${_$backColorAtom.name}_set');
  }


```

**Comment by pavanpodila** [comment]
This looks like a bug in the `analyzer` package. The visitor for field-element itself is returning `List<dynamic>` instead of `List<Color>`. @shyndman looks like we may need to see if a more recent version of analyzer fixes this. 


```dart
  void visitFieldElement(FieldElement element) { } // element is List<dynamic> !!!
```

Note that this is not a bug with `mobx_codgen`, but with the `analyzer` package. We will investigate and get back. Thanks so much for bringing to our attention!

**Comment by shyndman** [comment]
Do you have a local repro @pavanpodila?

@irvine5k, can you attach your entire `pubspec.lock`?

**Comment by pavanpodila** [comment]
@shyndman pls see this commit that shows the test that fails: https://github.com/mobxjs/mobx.dart/commit/2b46073346c1aaad40f3522351ca3de158485402

If you place you debug breakpoint for `visitFieldElement`, you can see the failing type as `List<dynamic>`

**Comment by irvine5k** [comment]
> Do you have a local repro @pavanpodila?
> 
> @irvine5k, can you attach your entire `pubspec.lock`?

```
# Generated by pub
# See https://dart.dev/tools/pub/glossary#lockfile
packages:
  _fe_analyzer_shared:
    dependency: transitive
    description:
      name: _fe_analyzer_shared
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.0.1"
  analyzer:
    dependency: transitive
    description:
      name: analyzer
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.39.2+1"
  archive:
    dependency: transitive
    description:
      name: archive
      url: "https://pub.dartlang.org"
    source: hosted
    version: "2.0.11"
  args:
    dependency: transitive
    description:
      name: args
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.5.2"
  async:
    dependency: transitive
    description:
      name: async
      url: "https://pub.dartlang.org"
    source: hosted
    version: "2.4.0"
  boolean_selector:
    dependency: transitive
    description:
      name: boolean_selector
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.0.5"
  build:
    dependency: transitive
    description:
      name: build
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.2.2"
  build_config:
    dependency: transitive
    description:
      name: build_config
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.4.1+1"
  build_daemon:
    dependency: transitive
    description:
      name: build_daemon
      url: "https://pub.dartlang.org"
    source: hosted
    version: "2.1.2"
  build_resolvers:
    dependency: transitive
    description:
      name: build_resolvers
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.3.0"
  build_runner:
    dependency: "direct dev"
    description:
      name: build_runner
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.7.3"
  build_runner_core:
    dependency: transitive
    description:
      name: build_runner_core
      url: "https://pub.dartlang.org"
    source: hosted
    version: "4.3.0"
  built_collection:
    dependency: transitive
    description:
      name: built_collection
      url: "https://pub.dartlang.org"
    source: hosted
    version: "4.3.2"
  built_value:
    dependency: transitive
    description:
      name: built_value
      url: "https://pub.dartlang.org"
    source: hosted
    version: "7.0.4"
  charcode:
    dependency: transitive
    description:
      name: charcode
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.1.2"
  checked_yaml:
    dependency: transitive
    description:
      name: checked_yaml
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.0.2"
  code_builder:
    dependency: transitive
    description:
      name: code_builder
      url: "https://pub.dartlang.org"
    source: hosted
    version: "3.2.1"
  collection:
    dependency: transitive
    description:
      name: collection
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.14.11"
  convert:
    dependency: transitive
    description:
      name: convert
      url: "https://pub.dartlang.org"
    source: hosted
    version: "2.1.1"
  crypto:
    dependency: transitive
    description:
      name: crypto
      url: "https://pub.dartlang.org"
    source: hosted
    version: "2.1.3"
  csslib:
    dependency: transitive
    description:
      name: csslib
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.16.1"
  cupertino_icons:
    dependency: "direct main"
    description:
      name: cupertino_icons
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.1.3"
  dart_style:
    dependency: transitive
    description:
      name: dart_style
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.3.3"
  fixnum:
    dependency: transitive
    description:
      name: fixnum
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.10.11"
  flutter:
    dependency: "direct main"
    description: flutter
    source: sdk
    version: "0.0.0"
  flutter_mobx:
    dependency: "direct main"
    description:
      name: flutter_mobx
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.3.6"
  flutter_test:
    dependency: "direct dev"
    description: flutter
    source: sdk
    version: "0.0.0"
  glob:
    dependency: transitive
    description:
      name: glob
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.2.0"
  graphs:
    dependency: transitive
    description:
      name: graphs
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.2.0"
  html:
    dependency: transitive
    description:
      name: html
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.14.0+3"
  http:
    dependency: transitive
    description:
      name: http
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.12.0+3"
  http_multi_server:
    dependency: transitive
    description:
      name: http_multi_server
      url: "https://pub.dartlang.org"
    source: hosted
    version: "2.1.0"
  http_parser:
    dependency: transitive
    description:
      name: http_parser
      url: "https://pub.dartlang.org"
    source: hosted
    version: "3.1.3"
  image:
    dependency: transitive
    description:
      name: image
      url: "https://pub.dartlang.org"
    source: hosted
    version: "2.1.4"
  io:
    dependency: transitive
    description:
      name: io
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.3.3"
  js:
    dependency: transitive
    description:
      name: js
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.6.1+1"
  json_annotation:
    dependency: transitive
    description:
      name: json_annotation
      url: "https://pub.dartlang.org"
    source: hosted
    version: "3.0.1"
  logging:
    dependency: transitive
    description:
      name: logging
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.11.3+2"
  matcher:
    dependency: transitive
    description:
      name: matcher
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.12.6"
  meta:
    dependency: transitive
    description:
      name: meta
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.1.8"
  mime:
    dependency: transitive
    description:
      name: mime
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.9.6+3"
  mobx:
    dependency: "direct main"
    description:
      name: mobx
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.4.0+1"
  mobx_codegen:
    dependency: "direct dev"
    description:
      name: mobx_codegen
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.4.0+1"
  nested:
    dependency: transitive
    description:
      name: nested
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.0.4"
  node_interop:
    dependency: transitive
    description:
      name: node_interop
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.0.3"
  node_io:
    dependency: transitive
    description:
      name: node_io
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.0.1+2"
  package_config:
    dependency: transitive
    description:
      name: package_config
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.1.0"
  package_resolver:
    dependency: transitive
    description:
      name: package_resolver
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.0.10"
  path:
    dependency: transitive
    description:
      name: path
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.6.4"
  pedantic:
    dependency: transitive
    description:
      name: pedantic
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.8.0+1"
  petitparser:
    dependency: transitive
    description:
      name: petitparser
      url: "https://pub.dartlang.org"
    source: hosted
    version: "2.4.0"
  pool:
    dependency: transitive
    description:
      name: pool
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.4.0"
  provider:
    dependency: "direct main"
    description:
      name: provider
      url: "https://pub.dartlang.org"
    source: hosted
    version: "4.0.1"
  pub_semver:
    dependency: transitive
    description:
      name: pub_semver
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.4.2"
  pubspec_parse:
    dependency: transitive
    description:
      name: pubspec_parse
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.1.5"
  quiver:
    dependency: transitive
    description:
      name: quiver
      url: "https://pub.dartlang.org"
    source: hosted
    version: "2.0.5"
  shelf:
    dependency: transitive
    description:
      name: shelf
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.7.5"
  shelf_web_socket:
    dependency: transitive
    description:
      name: shelf_web_socket
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.2.3"
  sky_engine:
    dependency: transitive
    description: flutter
    source: sdk
    version: "0.0.99"
  source_gen:
    dependency: transitive
    description:
      name: source_gen
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.9.4+7"
  source_span:
    dependency: transitive
    description:
      name: source_span
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.5.5"
  stack_trace:
    dependency: transitive
    description:
      name: stack_trace
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.9.3"
  stream_channel:
    dependency: transitive
    description:
      name: stream_channel
      url: "https://pub.dartlang.org"
    source: hosted
    version: "2.0.0"
  stream_transform:
    dependency: transitive
    description:
      name: stream_transform
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.0.20"
  string_scanner:
    dependency: transitive
    description:
      name: string_scanner
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.0.5"
  term_glyph:
    dependency: transitive
    description:
      name: term_glyph
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.1.0"
  test_api:
    dependency: transitive
    description:
      name: test_api
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.2.11"
  timing:
    dependency: transitive
    description:
      name: timing
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.1.1+2"
  typed_data:
    dependency: transitive
    description:
      name: typed_data
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.1.6"
  vector_math:
    dependency: transitive
    description:
      name: vector_math
      url: "https://pub.dartlang.org"
    source: hosted
    version: "2.0.8"
  watcher:
    dependency: transitive
    description:
      name: watcher
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.9.7+13"
  web_socket_channel:
    dependency: transitive
    description:
      name: web_socket_channel
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.1.0"
  xml:
    dependency: transitive
    description:
      name: xml
      url: "https://pub.dartlang.org"
    source: hosted
    version: "3.5.0"
  yaml:
    dependency: transitive
    description:
      name: yaml
      url: "https://pub.dartlang.org"
    source: hosted
    version: "2.2.0"
sdks:
  dart: ">=2.7.0 <3.0.0"
  flutter: ">=1.12.1"

```

**Comment by shyndman** [comment]
> @shyndman pls see this commit that shows the test that fails: [2b46073](https://github.com/mobxjs/mobx.dart/commit/2b46073346c1aaad40f3522351ca3de158485402)
> 
> If you place you debug breakpoint for `visitFieldElement`, you can see the failing type as `List<dynamic>`

@pavanpodila OK, so the way this stuff works isn't obvious. Your test isn't working for other reasons.

It isn't enough to include a Flutter dependency for `build` to understand `dart:ui` and Flutter types. The library actually has to be running inside of the Flutter platform. You can see that check here: https://github.com/dart-lang/build/blob/master/build_resolvers/lib/src/resolver.dart#L326

This is only true as a result of a `flutter test` or a `flutter packages pub run build_runner {command}`.

**Comment by shyndman** [comment]
Thanks for the lock @irvine5k. Can I get you to try something?

From your project directory, can you run:
`flutter clean && flutter packages pub run build_runner build --delete-conflicting-outputs`

**Comment by irvine5k** [comment]
> Thanks for the lock @irvine5k. Can I get you to try something?
> 
> From your project directory, can you run:
> `flutter clean && flutter packages pub run build_runner build --delete-conflicting-outputs`
> 
> After that, do you still get the dynamics?
Yes, I do.
```

@override
  List<dynamic> get backColor {
    _$backColorAtom.context.enforceReadPolicy(_$backColorAtom);
    _$backColorAtom.reportObserved();
    return super.backColor;
  }
```

**Comment by shyndman** [comment]
Hm. Bear with me.

@irvine5k, if you clone https://github.com/shyndman/mobx_list_color_repro, and run `flutter packages pub run build_runner build`, does `cartouche.g.dart` work properly there?

**Comment by irvine5k** [comment]
> Hm. Bear with me.
> 
> @irvine5k, I created a very simple project, containing only `cartouche.dart` and the necessary dependencies. My local version is generating code correctly.
> 
> Can you clone https://github.com/shyndman/mobx_list_color_repro, and run `flutter packages pub run build_runner build`? Does `cartouche.g.dart` get generated properly there?

@shyndman It does not work for me.

```
@override
  List get backColor {
    _$backColorAtom.context.enforceReadPolicy(_$backColorAtom);
    _$backColorAtom.reportObserved();
    return super.backColor;
  }

  @override
  set backColor(List value) {
    _$backColorAtom.context.conditionallyRunInAction(() {
      super.backColor = value;
      _$backColorAtom.reportChanged();
    }, _$backColorAtom, name: '${_$backColorAtom.name}_set');
  }
```

**Comment by shyndman** [comment]
hm, interesting that the dynamics are implicit here.

ok, next. what's your `flutter --version`?

**Comment by irvine5k** [comment]
@shyndman 

```
Flutter 1.12.13+hotfix.5 • channel stable • https://github.com/flutter/flutter.git
Framework • revision 27321ebbad (4 weeks ago) • 2019-12-10 18:15:01 -0800
Engine • revision 2994f7e1e6
Tools • Dart 2.7.0
```

**Comment by shyndman** [comment]
k, give me a few

**Comment by shyndman** [comment]
OK, I have a reproduction! Nice.

Give me a few more. :)

**Comment by shyndman** [comment]
Alright, so stable and beta both produce incorrect output.

I have verified that two options fix the output:
1. Upgrade to Flutter dev or master channels
2. Downgrade your mobx dependencies to:
   `mobx: ^0.3.6`
   `mobx_codegen: '0.3.11'`

@pavanpodila and I will introduce tests and a better process to make sure this doesn't happen again. Sorry about that.

**Comment by irvine5k** [comment]
No problem, @shyndman.

Thanks for your time.

**Comment by ricbermo** [comment]
To contribute a little bit. I was having an issue with `Lists`. I was getting a `List ins't a type` error which I solved by downgrading mobx and mobx_codege to the same versions.

**Comment by irvine5k** [comment]
> To contribute a little bit. I was having an issue with `Lists`. I was getting a `List ins't a type` error which I solved by downgrading mobx and mobx_codege to the same versions.

@ricbermo Can you explain more detailed your error and a way to @shyndman or @pavanpodila to reproduce it?

**Comment by ricbermo** [comment]
@irvine5k sure. I'm learning flutter and Mobx so I'm following this tutorial: https://mobx.pub/examples/github-repos. I was using the latest versions (screenshot1) and I was getting errors about `List`, `Future` not being a `type`.
The exact errors were:
1. `non_type_as_type_argument: The name 'List' isn't a type so it can't be used as a type argument`
2. `not_a_type: Future isn't a type`.

Here's the store:
```dart
import 'package:mobx/mobx.dart';
import 'package:github/github.dart';

part 'github_store.g.dart';

class GithubStore = _GithubStore with _$GithubStore;

abstract class _GithubStore with Store {
  var github = GitHub();

  static ObservableFuture<List<Repository>> emptyResponse = ObservableFuture.value([]);
  List<Repository> repositories = [];

  @observable
  String user = '';

  @observable

  @observable
  ObservableFuture<List<Repository>> fetchReposFuture = emptyResponse;

  @computed
  bool get hasResults =>
      fetchReposFuture != emptyResponse &&
      fetchReposFuture.status == FutureStatus.fulfilled;


  @action
  void setUser(String text) {
    print("hola");
    fetchReposFuture = emptyResponse;
    user = text;
  }

  @action
  Future<List<Repository>> fetchRepos() async {
    repositories = [];
    final future = github.repositories.listUserRepositories(user).toList();
    fetchReposFuture = ObservableFuture(future);

    return repositories = await future;
  }
}
```

After that, I decided to downgrade the mobx , flutter_mobx and mobx_codegen versions to `0.3.6`, `0.3.0` and `0.3.11` and the errors went away.

Let me know if you want me to add more details. TY.

Screenshot1: 
![image](https://user-images.githubusercontent.com/2601363/72088694-af15fd00-32d8-11ea-85d9-99c0fdf3307b.png)

**Comment by shyndman** [comment]
Hi @ricbermo,

Thanks. Yes, we encountered this issue in https://github.com/mobxjs/mobx.dart/issues/367. It was resolved by upgrading the `analyzer` package.

I'm going to be removing your comments shortly so the solution to this bug is clearer for other people encountering the issue.

**Comment by shyndman** [comment]
After a bit of research, I believe this to be the issue: https://github.com/dart-lang/build/issues/2528

**Comment by irvine5k** [comment]
@shyndman Any news?

This bug occurred with another classes and was reported in community, unfortunately I can't describe how to reproduce it, seems kinda random.

And it's not the first time that I see like it in the Flutterando community. I asked him(member of community that is struggling with this problem) to contribute in this issue but apparently he's not gonna to do it.

![image](https://user-images.githubusercontent.com/32758755/72907621-e772fd80-3d12-11ea-81dd-c1e66e3c34e9.png)

Maybe this "Category" is using dart:ui in its members but we need investigate if there's other classes with the same problem.

**Comment by shyndman** [comment]
Hi @irvine5k,

No progress on `build_resolvers` yet. I'm going to look soon.

Is this the `Category` class from the Flutter framework? If so, then yes, it's the same problem.

**Comment by irvine5k** [comment]
> Hi @irvine5k,
> 
> No progress on `build_resolvers` yet. I'm going to look soon.
> 
> Is this the `Category` class from the Flutter framework? If so, then yes, it's the same problem.

According him it's a class generated by Moor

![image](https://user-images.githubusercontent.com/32758755/72920161-8570c300-3d27-11ea-9e7e-60fa688c6abc.png)

**Comment by shyndman** [comment]
Thanks for the info.

I believe this is a whole other problem, and one that I've seen before — `build`'s inability to analyze the source code of another file generated via `build`.

That said, I'm not a `build` expert. There may be a way around this. I need to do some research.

**Comment by irvine5k** [comment]
> Thanks for the info.
> 
> I believe this is a whole other problem, and one that I've seen before — `build`'s inability to analyze the source code of another file generated via `build`. `mobx_codegen` can't actually _see_ the `Category` class, so it resolves to `dynamic`.
> 
> That said, I'm not a `build` expert. Let me verify that this is indeed the case, and if so, we can start looking into how to fix it.

Thanks for your response @shyndman and your attention for this issue. I'll reporting more bugs and helping with I can.

**Comment by shyndman** [comment]
https://github.com/dart-lang/build/pull/2603

**Comment by shyndman** [comment]
Pull merged. We're almost there.

Here's the `build_resolver` release PR.

**Comment by shyndman** [comment]
Alright, `build_resolvers` is published as version `1.3.2`. I'm putting together a PR right now to set it as our minimum version in `mobx_codegen`.

**Comment by irvine5k** [comment]
Great news o/
